Framework/API/src/Sample.cpp

```diff
@@ -146,8 +146,8 @@ const SampleEnvironment &Sample::getEnvironment() const {
  * @param env :: A pointer to a created sample environment. This takes
  * ownership of the object.
  */
-void Sample::setEnvironment(SampleEnvironment *env) {
-  m_environment = boost::shared_ptr<SampleEnvironment>(env);
+void Sample::setEnvironment(boost::shared_ptr<SampleEnvironment> env) {
```

Comment:
I'm not sure I like the change here which has resulted in `boost::shared_ptr` being specified as part of the public API. Internally, the `shared_ptr` does seem to be used to support cheap copy behaviour, but I see no reason why this needs to be leaked to the public API. i.e. the user of this type should not need to consider that the usage is shared (because as far as they are concerned it isn't). 

As in all cases, you are actually transferring ownership of the thing you make, I think it might make more sense to change this to be `Sample::setEnviroment(std::unique_ptr<SampleEnvironment> env)` and internally convert the unique_ptr to a shared_ptr. share_ptr already has a constructor overload to handle the creation of the shared_ptr for you.

Reply:
Ok, Ill look into making this change, I'm not entirely used to using smart pointers, but I know that we're meant to avoid using raw pointers wherever possible, so I simply changed that to stop it passing a raw pointer

Reply:
No worries. There are quite a few caveats to smart pointers. I wouldn't have expected that you would have known these already. Shared pointers are great for some things but often get used incorrectly and can then actually be quite hard to refactor out.

Reply:
so Just change the function to take a unique_ptr, then use that to create a shared_ptr within the function?

Reply:
Exactly. It sounds like a slightly crazy thing to do, but as I said above, the only reason `shared_ptr` is useful in this context for `Sample` is to allow cheap copies for `Sample`. It's not relevant to "users" of the `Shape` type.

Reply:
doing this seems to break LoadSampleEnvironment when it is updated to use unique_ptr instead of shared_ptr
```
std::unique_ptr<Geometry::SampleEnvironment> environment = nullptr;
  if (add) {
    environment = std::make_unique<Geometry::SampleEnvironment>(
        sample.getEnvironment());
    environment->add(environmentMesh);
  } else {
    auto can = boost::make_shared<Container>(environmentMesh);
    environment = std::make_unique<Geometry::SampleEnvironment>(name, can);
  }
  // Put Environment into sample.
  sample.setEnvironment(environment);
```
Which now fails due to a deleted function at `sample.setEnvironment(environment);`

Reply:
Ah I see, since its a unique pointer it has to be passed by reference to avoid copying

Reply:
It should be moved `std::move()`

Reply:
As in `sample.setEnvironment(std::move(environment)))`

Reply:
ah, I just changed it to a pass by reference, if the more correct thing to do in this scenario is to call std::move I'll change it now, thanks for the advice